src/Badges.sol

```diff
@@ -291,23 +321,46 @@ contract Badges is
     return tokenId;
   }
 
-  // signature comes from the people with the right params
+  // todo:: do we need to use a sparse markle tree? Is it more gas efficient?
+  // todo:: handling vouchers - what would be a voucherID here?
+  function safeCheckMerkleAgreement(
+    address _to,
+    address _from,
+    string calldata _uri,
+    bytes calldata _signature,
+    bytes32 _root,
+    bytes32[] calldata _proof
+  ) public view virtual returns (uint256) {
+    // this authenticates the signature coming from the issuer
+    bytes32 hash = getMerkleAgreementHash(_from, _uri, _root);
+    require(
+      SignatureCheckerUpgradeable.isValidSignatureNow(_from, hash, _signature),
+      "safeCheckMerkleAgreement: invalid signature"
+    );
+    bytes32 leaf = keccak256(abi.encodePacked(_to));
+    // this authenticates that the claimant is indeed part of the tree whose root was signed
+    require(MerkleProof.verify(_proof, _root, leaf), "safeCheckMerkleAgreement: invalid leaf");
+    uint256 voucherHashId = uint256(keccak256(abi.encode(_from, _to)));
+    return voucherHashId;
+  }
+
   function safeCheckAgreement(
     address _active,
     address _passive,
     string calldata _uri,
     bytes calldata _signature
-  ) internal virtual returns (uint256) {
+  ) internal view virtual returns (uint256) {
```

Comment:
Should we make this public?

Reply:
Hmm.. yeah. I think this then an internal operation. The safeCheckMerkleAgreement also should be internal I blv. Meaning the we need to test it with a different setup. Could you look into that, please?